### ISSUE TITLE
Logout/Restart/Shutdown popup-menu

### ISSUE THREAD

**Issue by franzmondlichtmann** [labwc/labwc#917]
I have the experience that I misclicked on the Exit button twice.
Now I have installed nwg-bar and replaced the default exit menu item with nwg-bar.
I think new users are turned off by the fact, that misclicking is so easy.
If a user is unlucky important work could be deleted.
Would it be a hassle to implement a Logout/Restart/Shutdown popup-menu?
Or using a existing open-source popup-menu?

**Comment by franzmondlichtmann** [comment]
Or at least asking the user if they are really sure?

**Comment by johanmalm** [comment]
Yes, we should do this really. One click to exit the compositor is too high risk.

I'm not sure what the best way to implement is - and welcome thoughts.

We could create a "Do you really want to exit?" OSD just like Openbox does.
Disadvantages:
- We have to translate the string
- We haven't done an OSD with buttons yet, so will add complexity
- Users are likely to want lots of config/theme options overtime

We could ship a dedicated "Do you really want to exit" client (possibly in python)
Disadvantages:
- Installation required (or at least copy to `~/bin`) before it can be used.
Advantages:
- Keeps `labwc` core codebase cleaner.
- Command line options could be added
- Could easily be extended to add "Reboot", "Shutdown", etc.

**Comment by Consolatis** [comment]
Too bad that there isn't a widely available wayland variant of `dialog` / `Xdialog`.

> * We have to translate the string

I don't think this is really an issue, if it isn't translated it would just show up in english as far as I understood the translation framework.

> We could ship a dedicated "Do you really want to exit" client (possibly in python)
> Disadvantages:

That external client would also require some kind of GUI framework, e.g. GTK or Qt and would thus make a default action depend on some GUI framework. Same thing for some dbus client that talks to a theoretical desktop portal interface (no clue if there is a generic "dialog" one or even a "shutdown" one which includes a "logout" option).

Kind of related:
- #892

**Comment by franzmondlichtmann** [comment]
I coudn't find an answer to this question:

> That external client would also require some kind of GUI framework, e.g. GTK or Qt and would thus make a default action depend on some GUI framework. Same thing for some dbus client that talks to a theoretical desktop portal interface (no clue if there is a generic "dialog" one or even a "shutdown" one which includes a "logout" option).

But here are some dbus commands for shutdown and logout.
I Don't know if they are working, I asked Bing AI.

Shutdown:
`dbus-send \
--system \
--dest=org.freedesktop.ConsoleKit \
--type=method_call \
--print-reply \
--reply-timeout=2000 \
/org/freedesktop/ConsoleKit/Manager \
org.freedesktop.ConsoleKit.Manager.Stop
`

Logout:
`
busctl call org.freedesktop.login1 /org/freedesktop/login1 \
org.freedesktop.login1.Manager TerminateSession s \
$(loginctl show-user $UID --property=Sessions --value)
`

**Comment by franzmondlichtmann** [comment]
Logout without busctl:
dbus-send --system --print-reply --dest=org.freedesktop.login1 \
/org/freedesktop/login1 'org.freedesktop.login1.Manager.TerminateSession' \
string:c2

**Comment by franzmondlichtmann** [comment]
I marked some of my comments, which were not very helpful, because they were x server related

**Comment by domo141** [comment]
FYI:
I currently use this (in a script that handles some other things, too):
```
if test "$1" = exit
then
        test "${LABWC_PID}" && dt="End LabWC" || dt="testidialooki"

        dt='<span size="32768">'$dt'</span>'
        zcmd='zenity --question --title=v'
        okl=--ok-label=t
        cnl=--cancel-label=nil
        $zcmd --text "$dt" $okl $cnl && kill $LABWC_PID || :
        exit 0
fi
```
works for me (although the zenity window appeared below this firefox one) for the time being...

**Comment by lobre** [comment]
I see there was this addition of `labnag` recently in the code base. I may be wrong, but it seems to be a built-in dialogue. If so, could it be useful for that issue?

https://github.com/labwc/labwc/commit/c63d35c9420bf0bdb7849dc00a2d679e9ac57524

And another related small question (even if not related), will `labnag` be usable outside of labwc? Say, for example, I want to implement a custom low-battery warning dialogue, which is aligned with my labwc theme. Do I have access to the `labnag` bin to invoke it?

**Comment by johanmalm** [comment]
> I see there was this addition of `labnag` recently in the code base. I may be wrong, but it seems to be a built-in dialogue. If so, could it be useful for that issue?

Yes, absolutely. Just use the `-Z` option.

> And another related small question (even if not related), will `labnag` be usable outside of labwc? Say, for example, I want to implement a custom low-battery warning dialogue, which is aligned with my labwc theme. Do I have access to the `labnag` bin to invoke it?

Yes.

**Comment by stefonarch** [comment]
Solution without buttons in niri

<img width="335" height="183" alt="Image" src="https://github.com/user-attachments/assets/2f458179-7c7b-48dd-9207-ae5bbd852509" />

**Comment by 01micko** [comment]
Here we go, I'll translate if you like too.

<details><summary>labexit</summary>

```sh
#!/bin/sh

# logout with labnag

labnag \
 -f "mono 10"\
 -m "Are you sure you want to quit Labwc?"\
 -Z " Yes " "labwc -e"\
 -Z " No  "\
 --background 00ffff\
 --button-background 00ffff\
 --border 00ccccaa\
 --text 000000\
 --button-text 000000\
 --button-gap 8\
 --button-margin-right 0\
 --button-padding 5\
 --button-border-size 2\
 -t 60
```

</details>

<details><summary>screenshot</summary>

<img width="1920" height="184" alt="Image" src="https://github.com/user-attachments/assets/7b3f5ff1-269b-4d42-b113-27e65368721f" />

</details>

Improvements welcome 😃

**Comment by 01micko** [comment]
Ok, I have it working nested with `LANGUAGE=it_IT.UTF-8`. Did a `DESTDIR` install and everything installs correctly. Quite a bit of mucking around to generate a good `labwc.pot` file so I'll have to update the docs about that.

<details><summary>screenhot running netsed with dodgy `it` translation</summary>

<img width="1283" height="750" alt="Image" src="https://github.com/user-attachments/assets/d85e0c49-e3d2-4a55-a5f4-7e6b36e4a2b0" />

</details>

**Comment by 01micko** [comment]
I think reboot, suspend, shutdown are out of scope for a simple `exit` trap. I did actually write one before this issue was raised for `labnag`. I'll post it for posterity. Of course different distros (gentoo, slackware come to mind) don't by default use `systemd`, although you can do what you like with either of those mentioned. The one below was written a few weeks ago when I was playing with `labnag` for slackware-current.

<details><summary>nagoff</summary>

```sh
#!/bin/sh

# logout with labnag

labnag \
 -f "Hack Regular 10"\
 -m "Choose your logout option"\
 -Z "  Lock   " "gtklock -d"\
 -Z " Logout  " "labwc -e"\
 -Z "Shutdown " "loginctl poweroff"\
 -Z " Reboot  " "loginctl reboot"\
 -Z "Hibernate" "loginctl hibernate"\
 -Z " Suspend " "loginctl suspend"\
 -Z " Cancel  "\
 --background 00ffff\
 --button-background 00ffff\
 --border 00ccccaa\
 --text 000000\
 --button-text 000000\
 --button-gap 8\
 --button-margin-right 0\
 --button-padding 5\
 --button-border-size 2\
 -t 60
```

</details>

`space` padding is used for consistent button size, but is not so important for a simple 2 button setup.
